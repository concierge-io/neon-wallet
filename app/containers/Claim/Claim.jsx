// @flow
import React, { Component } from 'react'
import ReactTooltip from 'react-tooltip'

type Props = {
  doClaimNotify: Function,
  setClaimRequest: Function,
  doGasClaim: Function,
  claimRequest: boolean,
  disableClaimButton: boolean,
  claimWasUpdated: boolean,
  claimAmount: number,
}

export default class Claim extends Component<Props> {
  componentDidUpdate () {
    const { claimRequest, claimWasUpdated, doClaimNotify, setClaimRequest } = this.props
    // if we requested a claim and new claims are available, do claim
    if (claimRequest && claimWasUpdated) {
      setClaimRequest(false)
      doClaimNotify()
    }
  }

  render () {
    const { claimAmount, disableClaimButton, doGasClaim } = this.props
    let renderButton
    if (!disableClaimButton) {
      renderButton = <button onClick={() => doGasClaim()}>Claim {claimAmount} GAS</button>
    } else {
      renderButton = (
        <div>
          <button data-tip data-for='claimTip' className='disabled'>Claim {claimAmount} GAS</button>
          <ReactTooltip class='solidTip' id='claimTip' place='bottom' type='dark' effect='solid'>
            <span>You can claim Gas once every 5 minutes</span>
          </ReactTooltip>
        </div>
      )
    }
    return <div id='claim'>{renderButton}</div>
  }
}
