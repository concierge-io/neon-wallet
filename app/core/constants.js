// @flow

export const NEON_WALLET_RELEASE_LINK = 'https://github.com/CityOfZion/neon-wallet/releases'

export const NETWORK = {
  MAIN: 'MainNet',
  TEST: 'TestNet',
  APHELION_MAIN: 'MainNet-A',
  APHELION_TEST: 'TestNet-A'
}

export const EXPLORER = {
  NEO_TRACKER: 'Neotracker',
  NEO_SCAN: 'Neoscan',
  ANT_CHAIN: 'Antchain'
}

export const ASSETS = {
  NEO: 'NEO',
  GAS: 'GAS'
}

export const ASSET_ID = {
  NEO: 'c56f33fc6ecfcd0c225c4ab356fee59390af8560be0e930faebe74a6daff7c9b',
  GAS: '602c79718b16e442de58778e148d0b1084e3b2dffd5de6b7b16cee7969282de7'
}

export const ASSETS_LABELS = {
  [ASSETS.NEO]: 'Neo',
  [ASSETS.GAS]: 'Gas'
}

export const DEFAULT_REQ = {
  jsonrpc: '',
  method: '',
  params: [],
  id: 0
}

export const ROUTES = {
  HOME: '',
  DASHBOARD: '/dashboard',
  CREATE_WALLET: '/create',
  ENCRYPT_KEY: '/encrypt-key',
  LOGIN_PRIVATE_KEY: '/login-private-key',
  LOGIN_LOCAL_STORAGE: '/login-local-storage',
  LOGIN_LEDGER_NANO_S: '/login-ledger-nano-s',
  LOGIN_NEP2: '/login-encrypted',
  LOGIN_TOKEN_SALE: '/login-token-sale',
  TOKEN_SALE: '/token-sale',
  SETTINGS: '/settings'
}

export const NOTIFICATION_LEVELS = {
  ERROR: 'error',
  SUCCESS: 'success',
  INFO: 'info',
  WARNING: 'warning'
}

export const NOTIFICATION_POSITIONS = {
  TOP_CENTER: 'tc',
  TOP_RIGHT: 'tr',
  TOP_LEFT: 'tl',
  BOTTOM_CENTER: 'bc',
  BOTTOM_RIGHT: 'br',
  BOTTOM_LEFT: 'bl'
}

export const BIP44_PATH =
  '8000002C' +
  '80000378' +
  '80000000' +
  '00000000' +
  '00000000'

export const MODAL_TYPES = {
  SEND_TRANSACTION: 'SEND_TRANSACTION',
  CONFIRM: 'CONFIRM'
}
