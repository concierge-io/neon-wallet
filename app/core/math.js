// @flow

// https://stackoverflow.com/questions/4912788/truncate-not-round-off-decimal-numbers-in-javascript
export const truncateNumber = (num: number, places: number): number =>
  Math.trunc(num * Math.pow(10, places)) / Math.pow(10, places)
