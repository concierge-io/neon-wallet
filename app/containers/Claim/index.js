// @flow
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import Claim from './Claim'
import {
  setClaimRequest,
  doGasClaim,
  doClaimNotify,
  getClaimAmount,
  getClaimRequest,
  getClaimWasUpdated,
  getDisableClaimButton
} from '../../modules/claim'

const mapStateToProps = (state: Object) => ({
  claimAmount: getClaimAmount(state),
  claimRequest: getClaimRequest(state),
  claimWasUpdated: getClaimWasUpdated(state),
  disableClaimButton: getDisableClaimButton(state)
})

const actionCreators = {
  setClaimRequest,
  doGasClaim,
  doClaimNotify
}

const mapDispatchToProps = dispatch => bindActionCreators(actionCreators, dispatch)

export default connect(mapStateToProps, mapDispatchToProps)(Claim)
