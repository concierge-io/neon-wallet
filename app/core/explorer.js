// @flow
import { openExternal } from './electron'
import { NETWORK, EXPLORER } from '../core/constants'

export const getExplorerLink = (net: NetworkType, explorer: ExplorerType, txid: string) => {
  let base
  if (explorer === EXPLORER.NEO_TRACKER) {
    if (net === NETWORK.MAIN) {
      base = 'https://neotracker.io/tx/'
    } else {
      base = 'https://testnet.neotracker.io/tx/'
    }
  } else if (explorer === EXPLORER.NEO_SCAN) {
    if (net === NETWORK.MAIN) {
      base = 'https://neoscan.io/transaction/'
    } else {
      base = 'https://neoscan-testnet.io/transaction/'
    }
  } else {
    if (net === NETWORK.MAIN) {
      base = 'http://antcha.in/tx/hash/'
    } else {
      base = 'http://testnet.antcha.in/tx/hash/'
    }
  }
  return `${base}${txid}`
}

export const openExplorer = (net: NetworkType, explorer: ExplorerType, txid: string) =>
  openExternal(getExplorerLink(net, explorer, txid))
