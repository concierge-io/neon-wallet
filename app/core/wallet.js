// @flow
import { ASSETS_LABELS } from './constants'
import { getAccountFromWIFKey, verifyAddress } from 'neon-js'

const MIN_PASSPHRASE_LEN = 4

export const validatePassphrase = (passphrase: string): boolean => passphrase.length >= MIN_PASSPHRASE_LEN

export const checkMatchingPassphrases = (passphrase: string, passphrase2: string) => passphrase !== passphrase2

export const verifyPrivateKey = (wif: string): boolean => {
  if (!wif) {
    return false
  }
  const account = getAccountFromWIFKey(wif)
  return account !== -1 && account.address
}

export const validateTransactionBeforeSending = (neoBalance: number, gasBalance: number, selectedAsset: string, selectedHash: string, balances: Object[], sendAddress: string, sendAmount: string) => {
  console.log('entered validation');
  if (!sendAddress || !sendAmount) {
    return {
      error: 'Please specify an address and amount',
      valid: false
    }
  }

  if (verifyAddress(sendAddress) !== true || sendAddress.charAt(0) !== 'A') {
    return {
      error: 'The address you entered was not valid.',
      valid: false
    }
  }

  if (selectedAsset === ASSETS_LABELS.NEO) {
    if (parseFloat(sendAmount) !== parseInt(sendAmount)) { // check for fractional neo
      return {
        error: 'You cannot send fractional amounts of Neo.',
        valid: false
      }
    }
    if (parseInt(sendAmount) > neoBalance) { // check for value greater than account balance
      return {
        error: 'You do not have enough NEO to send.',
        valid: false
      }
    }
  } else if (selectedAsset === ASSETS_LABELS.GAS) {
    if (parseFloat(sendAmount) > gasBalance) {
      return {
        error: 'You do not have enough GAS to send.',
        valid: false
      }
    }
  } else if(selectedHash){
    console.log('these are the balances', balances)
    if ( !balances[selectedHash] ||  parseFloat(sendAmount) > parseFloat(balances[selectedHash]) ){
      return {
        error: 'You do not have enough to send.',
        valid: false
      }
    }
  }

  if (parseFloat(sendAmount) < 0) { // check for negative asset
    return {
      error: 'You cannot send negative amounts of an asset.',
      valid: false
    }
  }

  return {
    error: '',
    valid: true
  }
}
