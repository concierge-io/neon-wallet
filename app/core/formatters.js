// @flow
import { truncateNumber } from './math'

const GAS_DECIMAL_LENGTH = 7
const GAS_DECIMAL_SHORT_DISPLAY_LENGTH = 4

const FIAT_DECIMAL_LENGTH = 2

export const formatGAS = (gas: number | string, shortDisplay: boolean = false): string => {
  const decimalLength = shortDisplay ? GAS_DECIMAL_SHORT_DISPLAY_LENGTH : GAS_DECIMAL_LENGTH
  return truncateNumber(parseFloat(gas), decimalLength).toFixed(decimalLength)
}

export const formatNEO = (neo: number | string): number => parseInt(neo, 10)

export const formatFiat = (value: number | string): string => parseFloat(value).toFixed(FIAT_DECIMAL_LENGTH)
