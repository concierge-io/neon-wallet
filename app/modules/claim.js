// @flow
import { doClaimAllGas, doSendAsset, getClaimAmounts, hardwareDoSendAsset, hardwareDoClaimAllGas } from 'neon-js'
import { log } from '../util/Logs'
import { ASSETS } from '../core/constants'
import { showErrorNotification, showSuccessNotification, showInfoNotification } from './notifications'
import { FIVE_MINUTES_MS } from '../core/time'
import { getWif, getAddress, getSigningFunction, getPublicKey, LOGOUT } from './account'
import { getNetwork } from './metadata'
import { getNeo } from './wallet'
import asyncWrap from '../core/asyncHelper'

// Constants
export const SET_CLAIM = 'SET_CLAIM'
export const SET_CLAIM_REQUEST = 'SET_CLAIM_REQUEST'
export const DISABLE_CLAIM = 'DISABLE_CLAIM'

// Actions
export function setClaim (claimAvailable: number, claimUnavailable: number) {
  return {
    type: SET_CLAIM,
    payload: {
      claimAvailable,
      claimUnavailable
    }
  }
}

export function setClaimRequest (claimRequest: boolean) {
  return {
    type: SET_CLAIM_REQUEST,
    payload: { claimRequest }
  }
}

export function disableClaim (disableClaimButton: boolean) {
  return {
    type: DISABLE_CLAIM,
    payload: { disableClaimButton }
  }
}

export const syncAvailableClaim = (net: NetworkType, address: string) => async (dispatch: DispatchType) => {
  const [_err, result] = await asyncWrap(getClaimAmounts(net, address)) // eslint-disable-line
  return dispatch(setClaim(result.available, result.unavailable))
}

export const doClaimNotify = () => async (dispatch: DispatchType, getState: GetStateType) => {
  const state = getState()
  const wif = getWif(state)
  const address = getAddress(state)
  const net = getNetwork(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)

  log(net, 'CLAIM', address, { info: 'claim all gas' })

  const isHardwareClaim = !!publicKey

  let claimGasFn
  if (isHardwareClaim) {
    dispatch(showInfoNotification({
      message: 'Sign transaction 2 of 2 to claim Gas on your hardware device (claiming Gas)',
      autoDismiss: 0
    }))
    claimGasFn = () => hardwareDoClaimAllGas(net, publicKey, signingFunction)
  } else {
    claimGasFn = () => doClaimAllGas(net, wif)
  }

  const [err, response] = await asyncWrap(claimGasFn())
  if (!err && response.result) {
    dispatch(showSuccessNotification({
      message: 'Claim was successful! Your balance will update once the blockchain has processed it.'
    }))
    setTimeout(() => dispatch(disableClaim(false)), FIVE_MINUTES_MS)
  } else {
    return dispatch(showErrorNotification({ message: 'Claim failed' }))
  }
}

// To initiate claim, first send all Neo to own address, the set claimRequest state
// When new claims are available, this will trigger the claim
export const doGasClaim = () => async (dispatch: DispatchType, getState: GetStateType) => {
  const state = getState()
  const wif = getWif(state)
  const address = getAddress(state)
  const net = getNetwork(state)
  const neo = getNeo(state)
  const signingFunction = getSigningFunction(state)
  const publicKey = getPublicKey(state)

  // if no neo in account, no need to send to self first
  if (neo === 0) {
    return dispatch(doClaimNotify())
  } else {
    dispatch(showInfoNotification({ message: 'Sending Neo to Yourself...', autoDismiss: 0 }))
    log(net, 'SEND', address, { to: address, amount: neo, asset: ASSETS.NEO })

    const isHardwareClaim = !!publicKey

    let sendAssetFn
    if (isHardwareClaim) {
      dispatch(showInfoNotification({
        message: 'Sign transaction 1 of 2 to claim Gas on your hardware device (sending Neo to yourself)',
        autoDismiss: 0
      }))
      sendAssetFn = () => hardwareDoSendAsset(net, address, publicKey, { [ASSETS.NEO]: neo }, signingFunction)
    } else {
      sendAssetFn = () => doSendAsset(net, address, wif, { [ASSETS.NEO]: neo })
    }

    const [err, response] = await asyncWrap(sendAssetFn())
    if (err || response.result === undefined || response.result === false) {
      return dispatch(showErrorNotification({ message: 'Transaction failed!' }))
    } else {
      dispatch(showInfoNotification({ message: 'Waiting for transaction to clear...', autoDismiss: 0 }))
      dispatch(setClaimRequest(true))
      return dispatch(disableClaim(true))
    }
  }
}

// State Getters
export const getClaimRequest = (state) => state.claim.claimRequest
export const getClaimAmount = (state) => state.claim.claimAmount
export const getClaimAvailable = (state) => state.claim.claimAvailable
export const getClaimUnavailable = (state) => state.claim.claimUnavailable
export const getClaimWasUpdated = (state) => state.claim.claimWasUpdated
export const getDisableClaimButton = (state) => state.claim.disableClaimButton

const initialState = {
  claimRequest: false,
  claimAmount: 0,
  claimAvailable: 0,
  claimUnavailable: 0,
  claimWasUpdated: false,
  disableClaimButton: false
}

export default (state: Object = initialState, action: Object) => {
  switch (action.type) {
    case SET_CLAIM_REQUEST:
      const { claimRequest } = action.payload
      return {
        ...state,
        claimRequest
      }
    case SET_CLAIM:
      const { claimAvailable, claimUnavailable } = action.payload
      let claimWasUpdated = false
      if (claimAvailable > state.claimAvailable && state.claimRequest === true) {
        claimWasUpdated = true
      }
      return {
        ...state,
        claimAmount: (claimAvailable + claimUnavailable) / 100000000,
        claimAvailable,
        claimUnavailable,
        claimWasUpdated
      }
    case DISABLE_CLAIM:
      const { disableClaimButton } = action.payload
      return {
        ...state,
        disableClaimButton
      }
    case LOGOUT:
      return initialState
    default:
      return state
  }
}
