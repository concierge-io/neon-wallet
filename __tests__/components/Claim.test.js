import React from 'react'
import { cloneDeep } from 'lodash'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { shallow, mount } from 'enzyme'
import Claim from '../../app/containers/Claim'
import * as neonjs from 'neon-js'
import { setClaimRequest, disableClaim } from '../../app/modules/claim'
import { SHOW_NOTIFICATION, HIDE_NOTIFICATIONS, DEFAULT_POSITION } from '../../app/modules/notifications'
import { NOTIFICATION_LEVELS } from '../../app/core/constants'

const initialState = {
  claim: {
    claimAmount: 10,
    claimRequest: false,
    claimWasUpdated: false,
    disableClaimButton: false
  },
  account: {
    wif: 'wif',
    address: 'address'
  },
  metadata: {
    network: 'network'
  },
  wallet: {
    Neo: 1
  }
}

const setup = (state = initialState, shallowRender = true) => {
  const store = configureStore([thunk])(state)

  let wrapper
  if (shallowRender) {
    wrapper = shallow(<Claim store={store} />)
  } else {
    wrapper = mount(
      <Provider store={store}>
        <Claim />
      </Provider>
    )
  }

  return {
    store,
    wrapper
  }
}

describe('Claim', () => {
  test('should render without crashing', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  test('should render claim gas button when claim button is not disabled', () => {
    const { wrapper } = setup()
    expect(wrapper.dive()).toMatchSnapshot()
  })

  test('should not render claim gas button when claim button is disabled', () => {
    const newState = cloneDeep(initialState)
    newState.claim.disableClaimButton = true
    const { wrapper } = setup(newState)
    expect(wrapper.dive()).toMatchSnapshot()
  })

  describe('when do gas claim button is clicked', () => {
    test('should dispatch transaction failure event', async () => {
      const { wrapper, store } = setup()
      neonjs.doSendAsset = jest.fn(() => {
        return new Promise((resolve, reject) => {
          resolve({ result: undefined })
        })
      })
      wrapper.dive().find('#claim button').simulate('click')

      await Promise.resolve().then().then().then()
      const actions = store.getActions()
      expect(actions.length).toEqual(4)
      expect(actions[0]).toEqual({
        type: HIDE_NOTIFICATIONS,
        payload: expect.objectContaining({
          dismissible: true,
          position: DEFAULT_POSITION
        })
      })
      expect(actions[1]).toEqual({
        type: SHOW_NOTIFICATION,
        payload: expect.objectContaining({
          message: 'Sending Neo to Yourself...',
          level: NOTIFICATION_LEVELS.INFO
        })
      })
      expect(actions[2]).toEqual({
        type: HIDE_NOTIFICATIONS,
        payload: expect.objectContaining({
          dismissible: true,
          position: DEFAULT_POSITION
        })
      })
      expect(actions[3]).toEqual({
        type: SHOW_NOTIFICATION,
        payload: expect.objectContaining({
          message: 'Transaction failed!',
          level: NOTIFICATION_LEVELS.ERROR
        })
      })
    })

    test('should dispatch transaction waiting, set claim request and disable claim event', async () => {
      const { wrapper, store } = setup()
      neonjs.doSendAsset = jest.fn(() => {
        return new Promise((resolve, reject) => {
          resolve({ result: true })
        })
      })

      wrapper.dive().find('#claim button').simulate('click')

      await Promise.resolve().then().then().then()
      const actions = store.getActions()
      expect(actions.length).toEqual(6)
      expect(actions[0]).toEqual({
        type: HIDE_NOTIFICATIONS,
        payload: expect.objectContaining({
          dismissible: true,
          position: DEFAULT_POSITION
        })
      })
      expect(actions[1]).toEqual({
        type: SHOW_NOTIFICATION,
        payload: expect.objectContaining({
          message: 'Sending Neo to Yourself...',
          level: NOTIFICATION_LEVELS.INFO
        })
      })

      expect(actions[2]).toEqual({
        type: HIDE_NOTIFICATIONS,
        payload: expect.objectContaining({
          dismissible: true,
          position: DEFAULT_POSITION
        })
      })
      expect(actions[3]).toEqual({
        type: SHOW_NOTIFICATION,
        payload: expect.objectContaining({
          message: 'Waiting for transaction to clear...',
          level: NOTIFICATION_LEVELS.INFO
        })
      })
      expect(actions[4]).toEqual(setClaimRequest(true))
      expect(actions[5]).toEqual(disableClaim(true))
    })
  })
})
