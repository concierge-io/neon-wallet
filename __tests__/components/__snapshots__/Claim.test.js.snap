// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Claim should not render claim gas button when claim button is disabled 1`] = `
<div
  id="claim"
>
  <div>
    <button
      className="disabled"
      data-for="claimTip"
      data-tip={true}
    >
      Claim 
      10
       GAS
    </button>
    <ReactTooltip
      class="solidTip"
      effect="solid"
      id="claimTip"
      insecure={true}
      place="bottom"
      resizeHide={true}
      type="dark"
      wrapper="div"
    >
      <span>
        You can claim Gas once every 5 minutes
      </span>
    </ReactTooltip>
  </div>
</div>
`;

exports[`Claim should render claim gas button when claim button is not disabled 1`] = `
<div
  id="claim"
>
  <button
    onClick={[Function]}
  >
    Claim 
    10
     GAS
  </button>
</div>
`;

exports[`Claim should render without crashing 1`] = `
<Claim
  claimAmount={10}
  claimRequest={false}
  claimWasUpdated={false}
  disableClaimButton={false}
  doClaimNotify={[Function]}
  doGasClaim={[Function]}
  setClaimRequest={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
